File: hash-table/4sum-ii.py
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        records = {}
        count = 0
        
        for num_1 in nums1:
            for num_2 in nums2:
                if num_1+num_2 in records:
                    records[num_1+num_2]+=1
                else:
                    records[num_1+num_2]=1

        for num_3 in nums3:
            for num_4 in nums4:
                if 0-num_3-num_4 in records:
                    count+=records[0-num_3-num_4]

        return count

File: hash-table/4sum-ii.cpp
class Solution {
public:
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) {
        unordered_map<int, int> records;
        int count = 0;

        for (int numA:nums1){
            for (int numB:nums2){
                records[numA+numB]++;
            }
        }

        for (int numC:nums3){
            for (int numD:nums4){ 
                int key = 0-numC-numD;
                if (records.find(key) != records.end()){
                    count+=records[key];
                }
            }
        }
        return count;
    }
};